### ISSUE TITLE
"error LNK2001: unresolved external symbol" when building _C.lib & _C.exp

### ISSUE THREAD

**Issue by qfdz01** [NVIDIAGameWorks/kaolin#482]
I encountered the fatal error when installing Kaolin using Anaconda. Everything work out just fine and ninja successfully built all the .obj files, but then the following error emerged:
```
正在创建库 C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\_C.lib 和对象C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\_C.exp
sided_distance.obj : error LNK2001: 无法解析的外部符号 "void __cdecl kaolin::sided_distance_forward_cuda_impl(class at::Tensor,class at::Tensor,class at::Tensor,class at::Tensor)" (?sided_distance_forward_cuda_impl@kaolin@@YAXVTensor@at@@000@Z)
build\lib.win-amd64-3.7\kaolin\_C.pyd : fatal error LNK1120: 1 个无法解析的外部命令
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX86\\x64\\link.exe' failed with exit status 1120
```
The whole message looks like this:
```
(kaolin) PS C:\Users\nians\Desktop\kaolin-master> python setup.py develop
Warning: passing language='c++' to cythonize() is deprecated. Instead, put "# distutils: language=c++" in your .pyx or .pxd file(s)
running develop
running egg_info
writing kaolin.egg-info\PKG-INFO
writing dependency_links to kaolin.egg-info\dependency_links.txt
writing requirements to kaolin.egg-info\requires.txt
writing top-level names to kaolin.egg-info\top_level.txt
reading manifest file 'kaolin.egg-info\SOURCES.txt'
reading manifest template 'MANIFEST.in'
adding license file 'LICENSE'
writing manifest file 'kaolin.egg-info\SOURCES.txt'
running build_ext
building 'kaolin._C' extension
Emitting ninja build file C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\build.ninja...
Compiling objects...
Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)
1.10.2.git.kitware.jobserver-1
C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:C:\Users\nians\Anaconda3\envs\kaolin\lib\site-packages\torch\lib "/LIBPATH:C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2\lib/x64" /LIBPATH:C:\Users\nians\Anaconda3\envs\kaolin\libs /LIBPATH:C:\Users\nians\Anaconda3\envs\kaolin\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\ATLMFC\lib\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\lib\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\um\x64" c10.lib torch.lib torch_cpu.lib torch_python.lib cudart_static.lib c10_cuda.lib torch_cuda.lib /EXPORT:PyInit__C C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\bindings.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\metrics\sided_distance.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\metrics\unbatched_triangle_distance.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\packed_simple_sum.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\tile_to_packed.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\conversions\mesh_to_spc\mesh_to_spc.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\conversions\unbatched_mcube\unbatched_mcube.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\mesh\mesh_intersection.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\convolution.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\feature_grids.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\point_utils.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\query.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\spc.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\render\dibr.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\render\spc\raytrace.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\metrics\sided_distance_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\metrics\unbatched_triangle_distance_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\packed_simple_sum_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\tile_to_packed_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\conversions\mesh_to_spc\mesh_to_spc_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\conversions\unbatched_mcube\unbatched_mcube_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\mesh\mesh_intersection_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\convolution_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\feature_grids_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\generate_points.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\minkowski_conv.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\point_utils_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\query_cuda.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\ops\spc\scan_octrees.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\render\dibr_back.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\render\dibr_for.obj C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\render\spc\raytrace_cuda.obj /OUT:build\lib.win-amd64-3.7\kaolin\_C.pyd /IMPLIB:C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\_C.lib
  正在创建库 C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\_C.lib 和对象 C:\Users\nians\Desktop\kaolin-master\build\temp.win-amd64-3.7\Release\kaolin/csrc\_C.exp
sided_distance.obj : error LNK2001: 无法解析的外部符号 "void __cdecl kaolin::sided_distance_forward_cuda_impl(class at::Tensor,class at::Tensor,class at::Tensor,class at::Tensor)" (?sided_distance_forward_cuda_impl@kaolin@@YAXVTensor@at@@000@Z)
build\lib.win-amd64-3.7\kaolin\_C.pyd : fatal error LNK1120: 1 个无法解析的外部命令
error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX86\\x64\\link.exe' failed with exit status 1120
(kaolin) PS C:\Users\nians\Desktop\kaolin-master>
```
The environment, with GTX1060:
```
torch                 1.6.0+cu101
torchvision           0.7.0+cu101
ninja                 1.10.2.3
Pillow                8.4.0
Cython                0.29.20
scipy                 1.7.3
CUDA                  10.2
visual studio         2017
```
It seems that something is missing or not linked, so "void __cdecl kaolin::sided_distance_forward_cuda_impl(class at::Tensor,class at::Tensor,class at::Tensor,class at::Tensor)" is not recognized? But I cannot find out what's missing.

One way to bypass the problem is using the cpu-version torch, but some key functions of Kaolin require CUDA, and will alert something like "xxx must be CUDA Tensor".

**Comment by tovacinni** [comment]
Thanks for your interest in Kaolin!

Which commit are you specifically trying to use with this build?

**Comment by qfdz01** [comment]
Sorry that I do not understand what you mean by "which commit"... If you mean the version of kaolin then I just downloaded the latest version here on github; If it's about c++ toolkit then it's 2017 version and MSVC=14.16.27023. I tried 2019 version(MSVC=14.29.30133) too but neither worked.
I installed all the required packages and ran "python setup.py develop", following the installation part of the doc. There is an ERROR called "LNK1181: cannot open input file"(seems no .obj file is compiled), but I cd into the "Release" folder, ran "ninja", went back to "kaolin-master" folder, ran "python setup.py develop" again and it was fixed. And then it comes the "LNK2001".
